lib/puppet/network/http/connection.rb

```diff
@@ -189,6 +190,42 @@ def request_with_redirects(request, options)
             request.each do |header, value|
               current_request[header] = value
             end
+          when 429, 503
+            retry_after = current_response['Retry-After']
+
+            if retry_after.nil?
+              # Set return value and advance to next loop.
+              response = current_response
+              next
+            end
+
+            retry_after = begin
+                            Integer(retry_after)
+                          rescue TypeError, ArgumentError
+                            begin
+                              DateTime.rfc2822(retry_after)
+                            rescue ArgumentError
+                              Puppet.err(_('Received a %{status_code} response from the server, but the Retry-After header value of "%{retry_after}" could not be converted to an integer or RFC 2822 Date.') %
+                                         {status_code: current_response.code,
+                                          retry_after: retry_after.inspect})
+
+                              response = current_response
+                              next
+                            end
+                          end
+
+            retry_sleep = case retry_after
+                          when DateTime
+                            (retry_after.to_time - DateTime.now.to_time).to_i
```

Comment:
Is `DateTime.rfc822` in the same time zone / format as `DateTime.now`? Do we have to worry about ensuring both are in UTC? Or will Ruby handle ensuring we have the correctly specified time span here?

Reply:
DateTime objects retain time zone info, which allows for a meaningful comparison when both are cast to a local UNIX timestamp via `to_time`. RFC 2822 and its successors mandate that zone info be provided with a date string, so there should be an issue with ambiguous dates --- the `rfc2822` parser will also reject date strings without zone info.

The spec tests for a Date value in the Retry-After header also set up a mixed time zone between the client and server to exercise this case.